pylint/utils/checkerstats.py

```diff
@@ -1,30 +1,353 @@
+else:
+    from typing_extensions import Literal, TypedDict
 
-if TYPE_CHECKING:
-    from typing import Counter  # typing.Counter added in Python 3.6.1
+
+class BadNames(TypedDict):
+    argument: int
+    attr: int
+    klass: int
+    class_attribute: int
+    class_const: int
+    const: int
+    inlinevar: int
+    function: int
+    method: int
+    module: int
+    variable: int
+
+
+class CodeTypeCount(TypedDict):
+    code: int
+    comment: int
+    docstring: int
+    empty: int
+    total: int
+
+
+class DuplicatedLines(TypedDict):
+    nb_duplicated_lines: int
+    percent_duplicated_lines: float
+
+
+class NodeCount(TypedDict):
+    function: int
+    klass: int
+    method: int
+    module: int
+
+
+class UndocumentedNodes(TypedDict):
+    function: int
+    klass: int
+    method: int
+    module: int
+
+
+class ModuleStats(TypedDict):
+    convention: int
+    error: int
+    fatal: int
+    info: int
+    refactor: int
+    statement: int
+    warning: int
+
+
+# pylint: disable-next=too-many-instance-attributes
+class CheckerStats:
+    def __init__(
+        self,
+        bad_names: Optional[BadNames] = None,
+        by_module: Optional[Dict[str, ModuleStats]] = None,
+        by_msg: Optional[Dict[str, int]] = None,
+        code_type_count: Optional[CodeTypeCount] = None,
+        dependencies: Optional[Dict[str, Set[str]]] = None,
+        duplicated_lines: Optional[DuplicatedLines] = None,
+        node_count: Optional[NodeCount] = None,
+        undocumented: Optional[UndocumentedNodes] = None,
+    ) -> None:
+        self.bad_names = bad_names or BadNames(
+            argument=0,
+            attr=0,
+            klass=0,
+            class_attribute=0,
+            class_const=0,
+            const=0,
+            inlinevar=0,
+            function=0,
+            method=0,
+            module=0,
+            variable=0,
+        )
+        self.by_module: Dict[str, ModuleStats] = by_module or {}
+        self.by_msg: Dict[str, int] = by_msg or {}
+        self.code_type_count = code_type_count or CodeTypeCount(
+            code=0, comment=0, docstring=0, empty=0, total=0
+        )
+
+        self.dependencies: Dict[str, Set[str]] = dependencies or {}
+        self.duplicated_lines = duplicated_lines or DuplicatedLines(
+            nb_duplicated_lines=0, percent_duplicated_lines=0.0
+        )
+        self.node_count = node_count or NodeCount(
+            function=0, klass=0, method=0, module=0
+        )
+        self.undocumented = undocumented or UndocumentedNodes(
+            function=0, klass=0, method=0, module=0
+        )
+
+        self.convention = 0
+        self.error = 0
+        self.fatal = 0
+        self.info = 0
+        self.refactor = 0
+        self.statement = 0
+        self.warning = 0
+
+        self.global_note = 0
+        self.nb_duplicated_lines = 0
+        self.percent_duplicated_lines = 0.0
+
+    def __str__(self) -> str:
+        final_string = ""
```

Comment:
Maybe a single fstring starting with `f"""` would be easier to maintain and think about here.

Reply:
We can't use `\n` in f-strings. We could do something like: 
```python
line_end = "\n"
final_string = f"{stat}{line_end} ..."
```
But I don't think that's much better. 
I thought about doing something with iterating over the fields/attributes. But don't think this is really feasible.

Reply:
I was thinking of a litteral line break:

```python
    return f"""{foo}
{bar}
... 
"""

Reply:
Fixed!